File: sdks/typescript/src/encryption.ts
// SPDX-License-Identifier: AGPL-3.0-only
//! Client-side payload encryption (ADR-0012) for the Quiver TypeScript SDK.
//
// Mirrors the reference envelope in `quiver_crypto::payload` byte-for-byte: a
// caller seals a JSON payload with a 256-bit key Quiver never sees, and the
// server stores and returns it as an opaque blob it cannot read. Sealing uses
// XChaCha20-Poly1305 (the audited `@stablelib/xchacha20poly1305`, which
// interoperates with the Rust and Python implementations) with a fresh random
// 192-bit nonce.
//
// This module lives at the `quiver-client/encryption` subpath so the core
// client stays dependency-free; `@stablelib/xchacha20poly1305` is an optional
// peer dependency you install only to use these helpers:
//
//     pnpm add @stablelib/xchacha20poly1305
//
// Keep fields server-filterable by leaving them in cleartext and merging the
// sealed envelope alongside them — `open` reads only the reserved key:
//
//     import { PayloadCipher } from "quiver-client/encryption";
//     const cipher = PayloadCipher.fromHex("…64 hex chars…");
//     const payload = { tier: "gold", ...cipher.seal({ ssn: "078-05-1120" }) };
//     // ... upsert `payload`; the server only ever sees ciphertext for `ssn`.
//     const secret = cipher.open(payload); // -> { ssn: "078-05-1120" }
//
// Use a dedicated key for payload encryption; never reuse your at-rest
// `QUIVER_ENCRYPTION_KEY`. The client owns the key — losing it means the data
// is unrecoverable.

import { XChaCha20Poly1305 } from "@stablelib/xchacha20poly1305";

/** The reserved payload key under which a sealed envelope is stored. */
export const ENVELOPE_KEY = "__quiver_enc__";

const VERSION = 1;
const ALG = "xchacha20poly1305";
const NONCE_LEN = 24;
const TAG_LEN = 16;
const AAD = new TextEncoder().encode("quiver/payload/v1");

/** An error sealing or opening a client-side payload envelope. */
export class PayloadError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "PayloadError";
  }
}

/** A client-held key for sealing and opening payload envelopes (ADR-0012). */
export class PayloadCipher {
  readonly #key: Uint8Array;

  private constructor(key: Uint8Array) {
    this.#key = key;
  }

  /** Build a cipher from a raw 256-bit (32-byte) key. */
  static fromBytes(key: Uint8Array): PayloadCipher {
    if (key.length !== 32) {
      throw new PayloadError(`payload key must be 32 bytes, got ${key.length}`);
    }
    return new PayloadCipher(Uint8Array.from(key));
  }

  /** Build a cipher from a 64-character hex-encoded 256-bit key. */
  static fromHex(hex: string): PayloadCipher {
    const clean = hex.trim();
    if (!/^[0-9a-fA-F]{64}$/.test(clean)) {
      throw new PayloadError(
        `payload key must be 64 hex characters, got ${clean.length}`,
      );
    }
    const key = new Uint8Array(32);
    for (let i = 0; i < 32; i++) {
      key[i] = Number.parseInt(clean.slice(i * 2, i * 2 + 2), 16);
    }
    return new PayloadCipher(key);
  }

  /** Seal `plaintext` into a one-key envelope `{ [ENVELOPE_KEY]: { … } }`. Each
   * call uses a fresh random nonce, so the same value seals to different
   * ciphertext. */
  seal(plaintext: unknown): Record<string, unknown> {
    const json = JSON.stringify(plaintext);
    if (json === undefined) {
      throw new PayloadError("cannot seal a value that is not JSON-serializable");
    }
    const nonce = randomBytes(NONCE_LEN);
    const ciphertext = new XChaCha20Poly1305(this.#key).seal(
      nonce,
      new TextEncoder().encode(json),
      AAD,
    );
    return {
      [ENVELOPE_KEY]: {
        v: VERSION,
        alg: ALG,
        n: toBase64(nonce),
        ct: toBase64(ciphertext),
      },
    };
  }

  /** Open an envelope sealed by {@link seal}, returning the plaintext. `sealed`
   * may carry cleartext sibling fields; only the reserved key is read. A wrong
   * key or any tampering throws {@link PayloadError}. */
  open(sealed: unknown): unknown {
    if (!isSealed(sealed)) {
      throw new PayloadError("payload is not a quiver-encrypted envelope");
    }
    const envelope = (sealed as Record<string, unknown>)[ENVELOPE_KEY];
    if (typeof envelope !== "object" || envelope === null) {
      throw new PayloadError("envelope is not an object");
    }
    const env = envelope as Record<string, unknown>;
    if (env.v !== VERSION) {
      throw new PayloadError(`unsupported envelope version: ${String(env.v)}`);
    }
    if (env.alg !== ALG) {
      throw new PayloadError(`unsupported envelope algorithm: ${String(env.alg)}`);
    }
    const nonce = decodeField(env, "n");
    if (nonce.length !== NONCE_LEN) {
      throw new PayloadError(`nonce is ${nonce.length} bytes, expected ${NONCE_LEN}`);
    }
    const ciphertext = decodeField(env, "ct");
    if (ciphertext.length < TAG_LEN) {
      throw new PayloadError(
        `ciphertext is ${ciphertext.length} bytes, shorter than the ${TAG_LEN}-byte tag`,
      );
    }
    const message = new XChaCha20Poly1305(this.#key).open(nonce, ciphertext, AAD);
    if (message === null) {
      throw new PayloadError("wrong key or tampered ciphertext");
    }
    return JSON.parse(new TextDecoder().decode(message));
  }
}

/** Whether `value` carries a Quiver payload envelope. */
export function isSealed(value: unknown): boolean {
  return typeof value === "object" && value !== null && ENVELOPE_KEY in value;
}

function decodeField(envelope: Record<string, unknown>, field: string): Uint8Array {
  const raw = envelope[field];
  if (typeof raw !== "string") {
    throw new PayloadError(`missing envelope field '${field}'`);
  }
  return fromBase64(raw);
}

function randomBytes(length: number): Uint8Array {
  const webcrypto = globalThis.crypto;
  if (!webcrypto?.getRandomValues) {
    throw new PayloadError("a Web Crypto getRandomValues implementation is required");
  }
  return webcrypto.getRandomValues(new Uint8Array(length));
}

function toBase64(bytes: Uint8Array): string {
  let binary = "";
  const chunk = 0x8000;
  for (let i = 0; i < bytes.length; i += chunk) {
    binary += String.fromCharCode(...bytes.subarray(i, i + chunk));
  }
  return btoa(binary);
}

function fromBase64(text: string): Uint8Array {
  const binary = atob(text);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

File: sdks/python/src/quiver/encryption.py
# SPDX-License-Identifier: AGPL-3.0-only
"""Client-side payload encryption (ADR-0012) for the Quiver Python SDK.

This mirrors the reference envelope in ``quiver_crypto::payload`` byte-for-byte:
a caller seals a JSON payload with a 256-bit key Quiver never sees, and the
server stores and returns it as an opaque blob it cannot read. Sealing uses
**XChaCha20-Poly1305** (the libsodium ``xchacha20poly1305_ietf`` construction,
which interoperates with the Rust reference) with a fresh random 192-bit nonce.

Requires the optional ``[encryption]`` extra (PyNaCl)::

    pip install quiver-client[encryption]

Keep fields server-filterable by leaving them in cleartext and merging the
sealed envelope alongside them — :meth:`PayloadCipher.open` reads only the
reserved key and ignores cleartext siblings::

    from quiver.encryption import PayloadCipher

    cipher = PayloadCipher.from_hex("…64 hex chars…")
    payload = {"tier": "gold", **cipher.seal({"ssn": "078-05-1120"})}
    # ... upsert `payload`; the server only ever sees ciphertext for `ssn`.
    secret = cipher.open(payload)          # -> {"ssn": "078-05-1120"}

**Use a dedicated key for payload encryption; never reuse your at-rest
``QUIVER_ENCRYPTION_KEY``.** The client owns the key — losing it means the data
is unrecoverable.
"""

from __future__ import annotations

import base64
import json
import os
from typing import Any

#: The reserved payload key under which a sealed envelope is stored.
ENVELOPE_KEY = "__quiver_enc__"

_VERSION = 1
_ALG = "xchacha20poly1305"
_NONCE_LEN = 24
_TAG_LEN = 16
_AAD = b"quiver/payload/v1"


class PayloadError(Exception):
    """Base class for client-side payload encryption errors."""


class NotEncryptedError(PayloadError):
    """The value carries no Quiver envelope (the reserved key is absent)."""


class MalformedEnvelopeError(PayloadError):
    """The envelope is structurally invalid or uses an unsupported version."""


class DecryptError(PayloadError):
    """Decryption failed: the wrong key or a tampered ciphertext."""


def _bindings() -> Any:
    """Import PyNaCl's libsodium bindings, with a clear error if it is missing."""
    try:
        from nacl import bindings
    except ImportError as exc:  # pragma: no cover - exercised only without the extra
        raise PayloadError(
            "client-side payload encryption requires the optional dependency PyNaCl; "
            "install it with: pip install quiver-client[encryption]"
        ) from exc
    return bindings


class PayloadCipher:
    """A client-held key for sealing and opening payload envelopes (ADR-0012)."""

    __slots__ = ("_key",)

    def __init__(self, key: bytes) -> None:
        if len(key) != 32:
            raise ValueError(f"payload key must be 32 bytes, got {len(key)}")
        self._key = bytes(key)

    @classmethod
    def from_hex(cls, hex_key: str) -> PayloadCipher:
        """Build a cipher from a 64-character hex-encoded 256-bit key."""
        try:
            key = bytes.fromhex(hex_key.strip())
        except ValueError as exc:
            raise ValueError(f"invalid payload key: {exc}") from exc
        return cls(key)

    def seal(self, plaintext: Any) -> dict[str, Any]:
        """Seal ``plaintext`` into a one-key envelope ``{ENVELOPE_KEY: {...}}``.

        Each call uses a fresh random nonce, so sealing the same value twice
        yields different ciphertext.
        """
        bindings = _bindings()
        message = json.dumps(
            plaintext, separators=(",", ":"), ensure_ascii=False
        ).encode("utf-8")
        nonce = os.urandom(_NONCE_LEN)
        ciphertext = bindings.crypto_aead_xchacha20poly1305_ietf_encrypt(
            message, _AAD, nonce, self._key
        )
        return {
            ENVELOPE_KEY: {
                "v": _VERSION,
                "alg": _ALG,
                "n": base64.b64encode(nonce).decode("ascii"),
                "ct": base64.b64encode(ciphertext).decode("ascii"),
            }
        }

    def open(self, sealed: Any) -> Any:
        """Open an envelope sealed by :meth:`seal`, returning the plaintext.

        ``sealed`` may carry cleartext sibling fields; only the reserved key is
        read. A wrong key or any tampering raises :class:`DecryptError`.
        """
        if not is_sealed(sealed):
            raise NotEncryptedError("payload is not a quiver-encrypted envelope")
        envelope = sealed[ENVELOPE_KEY]
        if not isinstance(envelope, dict):
            raise MalformedEnvelopeError("envelope is not an object")
        if envelope.get("v") != _VERSION:
            raise MalformedEnvelopeError(
                f"unsupported envelope version: {envelope.get('v')!r}"
            )
        if envelope.get("alg") != _ALG:
            raise MalformedEnvelopeError(
                f"unsupported envelope algorithm: {envelope.get('alg')!r}"
            )
        nonce = _decode_field(envelope, "n")
        if len(nonce) != _NONCE_LEN:
            raise MalformedEnvelopeError(
                f"nonce is {len(nonce)} bytes, expected {_NONCE_LEN}"
            )
        ciphertext = _decode_field(envelope, "ct")
        if len(ciphertext) < _TAG_LEN:
            raise MalformedEnvelopeError(
                f"ciphertext is {len(ciphertext)} bytes, shorter than the "
                f"{_TAG_LEN}-byte tag"
            )
        bindings = _bindings()
        from nacl.exceptions import CryptoError as NaClCryptoError

        try:
            message = bindings.crypto_aead_xchacha20poly1305_ietf_decrypt(
                ciphertext, _AAD, nonce, self._key
            )
        except NaClCryptoError as exc:
            raise DecryptError("wrong key or tampered ciphertext") from exc
        return json.loads(message)


def is_sealed(value: Any) -> bool:
    """Whether ``value`` carries a Quiver payload envelope."""
    return isinstance(value, dict) and ENVELOPE_KEY in value


def _decode_field(envelope: dict[str, Any], field: str) -> bytes:
    raw = envelope.get(field)
    if not isinstance(raw, str):
        raise MalformedEnvelopeError(f"missing envelope field {field!r}")
    try:
        return base64.b64decode(raw, validate=True)
    except (ValueError, TypeError) as exc:
        raise MalformedEnvelopeError(
            f"envelope field {field!r} is not base64: {exc}"
        ) from exc
